aiida_quantumespresso/workflows/pdos.py

```diff
@@ -0,0 +1,564 @@
+            'Both the `base_scf` and `base_nscf.pw.parent_folder` inputs were provided. The SCF calculation will '
+            'be run with the inputs provided in `base_scf` and the `base_nscf.pw.parent_folder` will be ignored.'
+        )
+    elif not 'base_scf' in inputs and not 'parent_folder' in inputs['base_nscf']['pw']:
+        return 'Specifying either the `base_scf` or `base_nscf.pw.parent_folder` input is required.'
+
+
+def validate_base_scf(value, _):
+    """Validate the base_scf parameters."""
+    parameters = value['pw']['parameters'].get_dict()
+    if parameters.get('CONTROL', {}).get('calculation', 'scf') != 'scf':
+        return '`CONTOL.calculation` in `base_scf.pw.parameters` is not set to `scf`.'
+
+
+def validate_base_nscf(value, _):
+    """Validate the base_nscf parameters."""
+    parameters = value['pw']['parameters'].get_dict()
+    if parameters.get('CONTROL', {}).get('calculation', 'scf') != 'nscf':
+        return '`CONTOL.calculation` in `base_nscf.pw.parameters` is not set to `nscf`.'
+    if parameters.get('SYSTEM', {}).get('occupations', None) != 'tetrahedra':
+        return '`SYSTEM.occupations` in `base_nscf.pw.parameters` is not set to `tetrahedra`.'
+
+
+def validate_dos_parameters(value, _):
+    """Validate DOS parameters.
+
+    - shared: Emin | Emax | DeltaE
+    - dos.x only: ngauss | degauss | bz_sum
+    - projwfc.x only: ngauss | degauss | pawproj | n_proj_boxes | irmin(3,n_proj_boxes) | irmax(3,n_proj_boxes)
+
+    """
+    jsonschema.validate(value.get_dict(), get_parameter_schema())
+
+
+def clean_calcjob_remote(node):
+    """Clean the remote directory of a ``CalcJobNode``."""
+    cleaned = False
+    try:
+        node.outputs.remote_folder._clean()  # pylint: disable=protected-access
+        cleaned = True
+    except (IOError, OSError, KeyError):
+        pass
+    return cleaned
+
+
+def clean_workchain_calcs(workchain):
+    """Clean all remote directories of a workchain's descendant calculations."""
+    cleaned_calcs = []
+
+    for called_descendant in workchain.called_descendants:
+        if isinstance(called_descendant, orm.CalcJobNode):
+            if clean_calcjob_remote(called_descendant):
+                cleaned_calcs.append(called_descendant.pk)
+
+    return cleaned_calcs
+
+
+PwBaseWorkChain = plugins.WorkflowFactory('quantumespresso.pw.base')
+DosCalculation = plugins.CalculationFactory('quantumespresso.dos')
+ProjwfcCalculation = plugins.CalculationFactory('quantumespresso.projwfc')
+
+
+class PdosWorkChain(ProtocolMixin, WorkChain):
+    """A WorkChain to compute Total & Partial Density of States of a structure, using Quantum Espresso."""
+
+    @classmethod
+    def define(cls, spec):
+        # type: (ProcessSpec) -> None
+        # yapf: disable
+        """Define the process specification."""
+        super().define(spec)
+        spec.input('structure')
+
+        spec.expose_inputs(
+            PwBaseWorkChain,
+            namespace='base_scf',
```

Comment:
I would call the namespace `scf` in line with the `PwRelax` and `PwBands`

Reply:
Aren't the namespaces called `base` and `base_scf` for the `PwRelaxWorkChain`? Perhaps we should rename the `scf` and `bands` namespaces for the `PwBandsWorkChain`  to `base_scf` and `base_bands`? We could also rename those of the `PwRelaxWorkChain`, but then having e.g. `relax` instead of `base` would be confusing since the `relax` namespace for the `PwBandsWorkChain` corresponds to the `PwRelaxWorkChain`. So maybe we can consistently use namespaces of the entry points:

* `PwBaseWorkChain`: `base_<calculation_type>`
* `PwRelaxWorkChain`: `relax` (possibly add relax_type if there are multiple relax types in the work chain)
*  `PwBandsWorkChain`: `bands`
*  etc...

Reply:
Fair point. I meant mostly the `PwBands` here as being a "real" higher level workflow, just like this `PdosWorkChain`. There we chose `scf` and `bands` (could have been `nscf`) which I think are succinct but still convey the entire message. The `PwRelaxWorkChain` is actually a bit of an exception. Ideally, I would have called it `relax` and `scf` for the two `PwBaseWorkChain` namespaces, but the weird thing is that when you then expose the `PwRelaxWorkChain` in a namespace called `relax`, you would have the namespace nesting `relax.relax`. So when looking at the `PwBandsWorkChain`, the `PwRelaxWorkChain` and the one of this PR, I think the relax is the exception and using the behavior of the bands one for the pdos one feels most natural. Also, this works nicely with the `dos` and `projwfc` namespaces. On a side note, they now directly expose the calcjob and I was going to suggest using a base workchain, except they don't exist yet and didn't want to put that effort on this PR. Also these calculations are unlikely to fail and need error handlers, but at some point it might become useful to expose the base workchain instead of the calcjob. Also then just keeping `dos` and `projwfc` would seem nicest to me.

Reply:
Thinking about this further, I still wonder if we shouldn't just use `base_<calculation_type>` consistently. In case a work chain wraps the `PwBandsWorkChain`, we again get a `bands.bands` namespace nesting. Is there a reason not to use `base_<calculation_type>`?

Reply:
Will stick with your suggestion for now, opened #662 to discuss the possibility of adding a consistent namespaces scheme.